### PR TITLE
Add bitmap-ray and sprite-ray collision functions

### PR DESCRIPTION
**Pull request by matt439**
# Description

There has been growing interest in using SplashKit to perform ray-tracing in both 2D and 3D environments. However, there is a lack of ray-based functions, particularly with detecting pixel-based collisions. I have added bitmap-ray and sprite-ray functions which will allow for more advanced ray-tracing, going beyond the primitive shapes.

### quad_from
This function constructs a quad around a passed-in segment. The width of the quad along the segment is also passed in as an argument.

### bitmap_quad_collision
I made use of a function which I previously developed and submitted in a previous PR: https://github.com/thoth-tech/splashkit-core/pull/83. Unlike in the previous PR, I did not include the 4 overloads in this PR.

### bitmap_bounding_circle bug fix
`bitmap_ray_collision` calls `bitmap_bounding_circle` which contains a bug. The bug was detected and fixed in a previous PR: https://github.com/thoth-tech/splashkit-core/pull/80. I copied the bug fix into this PR as it was necessary for accurate functioning of `bitmap_ray_collision`.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

I added two new tests to `sktest`: one in each of `test_geometry.cpp` and `test_sprites.cpp`. These real-time graphical tests allow the user to control a ray by using the arrow keys and mouse. There are bitmaps and sprites in their respective tests which indicate whether a collision has occurred.

## Testing Checklist

- [x] Tested with sktest

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

File: coresdk/src/coresdk/collisions.cpp [hunk 1]
```diff
@@ -274,6 +276,60 @@ namespace splashkit_lib
...
+        if ( not quads_intersect(q1, q) ) return false;
+        
+        return _step_through_pixels(rect.width, rect.height, translation_matrix(rect.x, rect.y), bmp->cell_w, bmp->cell_h, translation, [&] (int ax, int ay, int bx, int by)
+                                    {
+                                        return pixel_drawn_at_point(bmp, cell, bx, by) && point_in_quad(point_at(rect.x + ax, rect.y + ay), q);
+                                    });
+    }
+
+    bool bitmap_ray_collision(bitmap bmp, int cell, const matrix_2d& translation, const point_2d& origin, const vector_2d& heading)
+    {
```
Comment (NoahJCross) [inline coresdk/src/coresdk/collisions.cpp:298]: This should check for an invalid ptr to the bitmap

**Review by NoahJCross — CHANGES_REQUESTED** [review]
Nice updates to the circle calculations - using the diagonal distance gives a more precise radius than the previous MAX approach. Other than needing the INVALID_PTR check in bitmap_ray_collision, everything else looks good to go.

**Comment by matt439** [comment]
Added a check for invalid bitmap pointer in `bitmap_ray_collision`.

**Review by NoahJCross — APPROVED** [review]
I've reviewed this PR and approve of these changes.
The ray collision functions add important features to SplashKit while keeping high standards:
# Code Quality
Good use of matrix transformations for bitmap calculations
Proper collision detection with bitmap pixels
Good handling of edge cases (null checks, out of bounds)
Clean code that fits well with existing collision systems
# Testing
Good test coverage in test_geometry.cpp
# Documentation
Clear methods for both bitmap and sprite collision checks
Consistent design with other collision functions
Well-documented code that explains the parameters

The changes add useful collision features while keeping types consistent and calculations accurate.

**Review by WhyPenguins — APPROVED** [review]
Looks good! I think implementing the ray collision via a new general case `bitmap_quad_collision` was a good choice. I think the only downside might be inefficiency when the ray is diagonal (I believe it has to check the entire sprite at that point?).

Tests look good and it works well! Nice work!

Just leaving a note here for anyone looking, future work could be to write a more special case ray-intersection function that might be quicker (probably via DDA raymarching), and also returning the hit point/distance. In the current setup I think this could be found by updating the hit point/distance inside the `_step_through_pixels` lambda, before returning the result.

**Comment by matt439** [comment]
Thanks for the feedback. Yes the edge case when the ray's heading matches the division between the triangles is a problem. Hopefully it does not occur too often but an improved solution would be a good task for next trimester.
